Aplib.Core/Belief/MemoryBelief.cs

```diff
@@ -0,0 +1,100 @@
+        /// and a function to generate/update the observation using the object reference.
+        /// Also initializes the memory array with a specified number of slots.
+        /// </summary>
+        /// <param name="reference">The reference used to generate/update the observation.</param>
+        /// <param name="getObservationFromReference">A function that takes a reference and generates/updates a observation.</param>
+        /// <param name="framesToRemember">The number of frames to remember back.</param>
+        public MemoryBelief(TReference reference, Func<TReference, TObservation> getObservationFromReference, int framesToRemember)
+            : base(reference, getObservationFromReference)
+        {
+            _memorizedObservations = new(framesToRemember);
+        }
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="MemoryBelief{TReference, TObservation}"/> class with an object reference,
+        /// a function to generate/update the observation using the object reference,
+        /// and a condition on when the observation should be updated.
+        /// Also initializes the memory array with a specified number of slots.
+        /// </summary>
+        /// <param name="reference">The reference used to generate/update the observation.</param>
+        /// <param name="getObservationFromReference">A function that takes a reference and generates/updates a observation.</param>
+        /// <param name="framesToRemember">The number of frames to remember back.</param>
+        /// <param name="shouldUpdate">A function that sets a condition on when the observation should be updated.</param>
+        public MemoryBelief(TReference reference, Func<TReference, TObservation> getObservationFromReference, int framesToRemember,
+            Func<bool> shouldUpdate)
+            : base(reference, getObservationFromReference, shouldUpdate)
+        {
+            _memorizedObservations = new(framesToRemember);
+        }
+
+        /// <summary>
+        /// Generates/updates the observation.
+        /// Also stores the previous observation in memory.
+        /// </summary>
+        public override void UpdateBelief()
+        {
+            // We use the implicit conversion to TObservation to store the observation
+            _memorizedObservations.Put(this);
+            base.UpdateBelief();
+        }
+
+        /// <summary>
+        /// Gets the most recently memorized observation.
+        /// </summary>
+        /// <returns> The most recent memory of the observation.</returns>
+        public TObservation GetMostRecentMemory() => _memorizedObservations.GetFirst();
+
+        /// <summary>
+        /// Gets the memorized observation at a specific index.
+        /// A higher index means a memory further back in time.
+        /// If the index is out of bounds, returns the closest element that is in bounds.
```

Comment:
I assume this means the closest element that is in bounds, closest to the specified index? It could also be interpreted as the closest element to the current frame/now/0 so specifying this might be nice.